Apache Flink connector docs (#5527)
ClickHouse/clickhouse-docs: docs/integrations/data-ingestion/apache-flink/flink-connector.md

        MAX_TIME_IN_BUFFER_MS, 
        MAX_RECORD_SIZE_IN_BYTES, 
        clickHouseClientConfig
);

csvSink.setClickHouseFormat(ClickHouseFormat.CSV);
```

3. Finally, connect your DataStream to the sink.

```java
data.sinkTo(csvSink);
```

More examples and snippets can be found in our tests:
- [flink-connector-clickhouse-1.17](https://github.com/ClickHouse/flink-connector-clickhouse/tree/main/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink)
- [flink-connector-clickhouse-2.0.0](https://github.com/ClickHouse/flink-connector-clickhouse/tree/main/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink)

### Quick Start Example {#datastream-quick-start}

We have created maven-based example for an easy start with the ClickHouse Sink:

- [Flink 1.17+](https://github.com/ClickHouse/flink-connector-clickhouse/tree/main/examples/maven/flink-v1.7/covid)
- [Flink 2.0.0+](https://github.com/ClickHouse/flink-connector-clickhouse/tree/main/examples/maven/flink-v2/covid)

For more detailed instructions, see the [Example Guide](https://github.com/ClickHouse/flink-connector-clickhouse/blob/main/examples/README.md)

> ### DataStream API Connection Options {#datastream-api-connection-options}

Author: addressed by the previous comment

> ### DataStream API Connection Options {#datastream-api-connection-options}

Reviewer: Tnx. Can you make it more explicit? 
There are two types of settings users can tune:
* Java client options https://clickhouse.com/docs/integrations/language-clients/java/client#configuration
* ClickHouse session settings https://clickhouse.com/docs/operations/settings/settings

With the existing phrasing, the only indication forthe clickhouse settings is by this line
`Map.of(), // serverSettings`.

| `java.util.UUID`    | `UUID`            | ✅         | `DataWriter.writeIntUUID`     |
| `String`            | `JSON`            | ✅         | `DataWriter.writeJSON`        |
| `Array<Type>`       | `Array<Type>`     | ✅         | `DataWriter.writeArray`       |
| `Map<K,V>`          | `Map<K,V>`        | ✅         | `DataWriter.writeMap`         |
| `Tuple<Type,..>`    | `Tuple<T1,T2,..>` | ✅         | `DataWriter.writeTuple`       |
| `Object`            | `Variant`         | ❌         | N/A                           |

Notes:
* A `ZoneId` must be provided when performing date operations.
* Precision and scale must be provided when performing decimal operations.
* In order for ClickHouse to parse a Java String as JSON, you need to enable `enableJsonSupportAsString` in `ClickHouseClientConfig`.

## Metrics {#metrics}

The connector exposes the following additional metrics on top of Flink's existing metrics:

| Metric                                  | Description                                                         | Type      | Status |
|-----------------------------------------|---------------------------------------------------------------------|-----------|--------|

> | `numBytesSend`                          | Total number of bytes sent to ClickHouse                            | Counter   | ✅      |

Author: good q - in a small test i ran, i see they are different. I don't exactly know why yet, but i'll note it down to look into more deeply.

> | `numBytesSend`                          | Total number of bytes sent to ClickHouse                            | Counter   | ✅      |

Reviewer 2: This is the size of data that is sent to ClickHouse, aka payload `RowBinary` or `RowBinaryWithDefaults`, not related to `written_bytes`

| `writeLatencyHistogram`                 | Histogram of successful write latency distribution (ms)             | Histogram | ✅      |
| `writeFailureLatencyHistogram`          | Histogram of failed write latency distribution (ms)                 | Histogram | ✅      |
| `triggeredByMaxBatchSizeCounter`        | Total number of flushes triggered by reaching `maxBatchSize`        | Counter   | ✅      |
| `triggeredByMaxBatchSizeInBytesCounter` | Total number of flushes triggered by reaching `maxBatchSizeInBytes` | Counter   | ✅      |
| `triggeredByMaxTimeInBufferMSCounter`   | Total number of flushes triggered by reaching `maxTimeInBufferMS`   | Counter   | ✅      |
| `actualRecordsPerBatch`                 | Histogram of actual batch size distribution                         | Histogram | ✅      |
| `actualBytesPerBatch`                   | Histogram of actual bytes per batch distribution                    | Histogram | ✅      |

[//]: # (| actualTimeInBuffer           | Histogram of actual time in buffer before flush distribution       | Histogram | ❌      |)

## Limitations {#limitations}

> * Currently, the sink does not support exactly-once semantics.

Reviewer: In reference to what @mshustov said earlier, let's make sure we have a point and a linked issue for each of these:
* Exactly once semantics
* DLQ
* TableAPI

A link for each one of these will allow the community/user to vote and give us feedback

| `password` | ClickHouse database password   | N/A           | Yes      |
| `database` | ClickHouse database name       | N/A           | Yes      |
| `table`    | ClickHouse table name          | N/A           | Yes      |

Additional options can be passed to the client as a `Map<String, String>`.
All available client options are listed in [ClientConfigProperties.java](https://github.com/ClickHouse/clickhouse-java/blob/main/client-v2/src/main/java/com/clickhouse/client/api/ClientConfigProperties.java).
For example:

<Tabs groupId="client_options_example">
<TabItem value="Java" label="Java" default>

```java
Map<String, String> additionalProperties = Map.of(
    ClientConfigProperties.CA_CERTIFICATE.getKey(), "<my_CA_cert>",
    ClientConfigProperties.SERVER_TIMEZONE.getKey(), "America/Los_Angeles"
);

ClickHouseClientConfig clickHouseClientConfig = new ClickHouseClientConfig(
    url,
    username,
    password,
    database,
    tableName,
    additionalProperties,
    Map.of(), // serverSettings
    false // enableJsonSupportAsString
);
```
</TabItem>
</Tabs>

#### Sink Options {#sink-options}

The following options come directly from Flink's `AsyncSinkBase`:

> | Parameters             | Description                                                                                                 | Default Value |

Reviewer: All options' default values are marked with N/A. If there truly are no defaults and users must always provide them, it would be worth adding a note saying these are all required (like was done in the client options section

Author: > it would be worth adding a note saying these are all required

done

| `maxTimeInBufferMS`    | The maximum time a record may stay in the sink before being flushed                                         | N/A           |
| `maxRecordSizeInBytes` | The maximum record size that the sink will accept, records larger than this will be automatically rejected  | N/A           |

## Supported data types {#supported-data-types}

The table below provides a quick reference for converting data types when inserting from Flink into ClickHouse.

### Inserting data from Flink into ClickHouse {#inserting-data-from-flink-into-clickhouse}

[//]: # (TODO: add a "Flink SQL Type" column once table api support is added )

| Java Type           | ClickHouse Type   | Supported | Serialize Method              |
|---------------------|-------------------|-----------|-------------------------------| 
| `byte`/`Byte`       | `Int8`            | ✅         | `DataWriter.writeInt8`        |
| `short`/`Short`     | `Int16`           | ✅         | `DataWriter.writeInt16`       |
| `int`/`Integer`     | `Int32`           | ✅         | `DataWriter.writeInt32`       |
| `long`/`Long`       | `Int64`           | ✅         | `DataWriter.writeInt64`       |

> | `BigInteger`        | `Int128`          | ✅         | `DataWriter.writeInt124`      |

Reviewer: ```suggestion
| `BigInteger`        | `Int128`          | ✅         | `DataWriter.writeInt128`      |
```

| `triggeredByMaxBatchSizeInBytesCounter` | Total number of flushes triggered by reaching `maxBatchSizeInBytes` | Counter   | ✅      |
| `triggeredByMaxTimeInBufferMSCounter`   | Total number of flushes triggered by reaching `maxTimeInBufferMS`   | Counter   | ✅      |
| `actualRecordsPerBatch`                 | Histogram of actual batch size distribution                         | Histogram | ✅      |
| `actualBytesPerBatch`                   | Histogram of actual bytes per batch distribution                    | Histogram | ✅      |

[//]: # (| actualTimeInBuffer           | Histogram of actual time in buffer before flush distribution       | Histogram | ❌      |)

## Limitations {#limitations}

* The sink does not support exactly-once delivery - it currently only provides an at-least-once guarantee. This feature is tracked [here](https://github.com/ClickHouse/flink-connector-clickhouse/issues/106).
* The sink does not support a dead-letter queue (DLQ) to buffer unprocessed messages. Currently, the connector will fail at the first row it cannot process. This feature is tracked [here](https://github.com/ClickHouse/flink-connector-clickhouse/issues/105).

> * The sink does not support creation via Flink's Table API/Flink SQL. This feature is tracked [here](https://github.com/ClickHouse/flink-connector-clickhouse/issues/42).

Reviewer: I'll suggest a softer version:

```suggestion
## Limitations {#limitations}

* The sink currently provides an at-least-once delivery guarantee. Work toward exactly-once semantics is being tracked [here](https://github.com/ClickHouse/flink-connector-clickhouse/issues/106).
* The sink does not yet support a dead-letter queue (DLQ) for buffering unprocessable messages. In the meantime, the connector will stop processing at the first row it is unable to handle. This feature is being tracked [here](https://github.com/ClickHouse/flink-connector-clickhouse/issues/105).
* The sink does not yet support creation via Flink's Table API or Flink SQL. This feature is being tracked [here](https://github.com/ClickHouse/flink-connector-clickhouse/issues/42).
```

## ClickHouse Version Compatibility and Security {#compatibility-and-security}

> - All artifacts and versions of the connector are tested with all [active and LTS versions](https://github.com/ClickHouse/ClickHouse/pulls?q=is%3Aopen+is%3Apr+label%3Arelease) of ClickHouse.

Reviewer: Given that the CH versions are static in our workflows and not yet fully automated to be updated, we should consider being more accurate, maybe something like:

```suggestion
- The connector is tested against a range of recent ClickHouse versions, including latest and head, via a daily CI workflow. The tested versions are updated periodically as new ClickHouse releases become active.
```

Adding a link to the matrix might be useful as well, so users would get an accurate status. https://github.com/ClickHouse/flink-connector-clickhouse/blob/main/.github/workflows/tests-nightly.yaml#L15

Author: thank you, i've updated this line
